src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -38,17 +43,44 @@ internal static class DependentTypeFinder
         private static readonly ObjectPool<SymbolAndProjectIdSet> s_setPool = new ObjectPool<SymbolAndProjectIdSet>(
             () => new SymbolAndProjectIdSet(SymbolAndProjectIdComparer<INamedTypeSymbol>.SymbolEquivalenceInstance));
 
+        // Caches from a types to their related types (in the context of a specific solution).
+        // Kept as a cache so that clients who make many calls into us won't end up computing
+        // the same data over and over again.  Will be let go the moment the solution they're
+        // based off of is no longer alive.
+
+        private static readonly RelatedTypeCache s_typeToImmediatelyDerivedClassesMap = new RelatedTypeCache();
+        private static readonly RelatedTypeCache s_typeToTransitivelyDerivedClassesMap = new RelatedTypeCache();
+        private static readonly RelatedTypeCache s_typeToTransitivelyImplementingTypesMap = new RelatedTypeCache();
+        private static readonly RelatedTypeCache s_typeToImmediatelyDerivedAndImplementingTypesMap = new RelatedTypeCache();
```

Comment:
Memory impact could be a concern.  Effectively if the solution is alive, and we've done some SymbolFinding operators, we could have some caches from types->related-types.  As a related type could be a source type, that means it's source-assembly would be held alive.  If the source-assembly is alive, it means its compilaiton is alive.

So, this could effectively keep the compilations for a solution alive as long as the solution itself was alive.

--

I think i'd prefer to probably take this for 15.6 so we can see if this is an issue.  It could also come online as we turn OOP on, as it would be nicer to have thse caches OOP instead of in the VS process space.